### ISSUE TITLE
Issue with Angular 2 and multiple skulpt files in a single code window.

### ISSUE THREAD

**Issue by viztastic** [skulpt/skulpt#603]
We have integrated Skulpt with an Angular 2 build and are using a single instance of code window and subsequently, the canvas element with Turtle draw. We are loading multiple files into the single code window and we have found that when a canvas drawing is currently being drawn and we attempt to load another file, the browser memory is exhausted and the window crashes.

Is this a known issue with Skulpt and if so, are there any workarounds you are familiar with or is there a fix already being worked on to support the cancellation of existing drawings prior to loading a new file in a Skulpt code window?

**Comment by waywaaard** [comment]
What do you mean by multiple skulpt files and how do you load a new file while the drawing is in process?

Can you break it down to a minimal example on jsFiddle maybe?

**Comment by viztastic** [comment]
Hi @waywaaard 

I just managed to get an example uploaded to show you what I'mt alking about:

When a user 'runs' file 3 and then attempts to run file 4 (while file 3 is still running), the screen will crash.

Link to app [here](http://skulpttest.s3-website-ap-southeast-2.amazonaws.com)

![image](https://cloud.githubusercontent.com/assets/8556736/17761301/4789bd44-6549-11e6-9ef7-cbdef6153d7a.png)

**Comment by waywaaard** [comment]
Yes, that is a fundamental problem of the skulpt architecture. There are a lot of things stored within the `Sk.*` namespace for running programs. So if you have 2 running programs at the same time it might crash.

You need to assure that the other program has already been terminated.

**Comment by viztastic** [comment]
@waywaaard Is there a recommended way of enforcing the "reassurance" ?

**Comment by waywaaard** [comment]
If you are using something like this:
```
    Sk.misceval.asyncToPromise(function() {
      return Sk.importMainWithBody("<stdin>",false,prog,true);
    });
```
Then you can add a `then` and a `catch` handler. If either then or catch is called, then skulpt has finished the current program.
You could use a flag for that or use wrap the promise so that you can have access to the internal state. Or wrap it if bluebird.
